iothub/device/samples/convention-based-samples/TemperatureController/TemperatureControllerSample.cs

```diff
@@ -0,0 +1,396 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    public class TemperatureControllerSample
+    {
+        private const string Thermostat1 = "thermostat1";
+        private const string Thermostat2 = "thermostat2";
+
+        private static readonly Random s_random = new Random();
+        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(5);
+
+        private readonly DeviceClient _deviceClient;
+        private readonly ILogger _logger;
+
+        // Dictionary to hold the temperature updates sent over each "Thermostat" component.
+        // NOTE: Memory constrained devices should leverage storage capabilities of an external service to store this
+        // information and perform computation.
+        // See https://docs.microsoft.com/en-us/azure/event-grid/compare-messaging-services for more details.
+        private readonly Dictionary<string, Dictionary<DateTimeOffset, double>> _temperatureReadingsDateTimeOffset =
+            new Dictionary<string, Dictionary<DateTimeOffset, double>>();
+
+        // Dictionary to hold the current temperature for each "Thermostat" component.
+        private readonly Dictionary<string, double> _temperature = new Dictionary<string, double>();
+
+        // Dictionary to hold the max temperature since last reboot, for each "Thermostat" component.
+        private readonly Dictionary<string, double> _maxTemp = new Dictionary<string, double>();
+
+        public TemperatureControllerSample(DeviceClient deviceClient, ILogger logger)
+        {
+            _deviceClient = deviceClient ?? throw new ArgumentNullException($"{nameof(deviceClient)} cannot be null.");
+            _logger = logger ?? LoggerFactory.Create(builer => builer.AddConsole()).CreateLogger<TemperatureControllerSample>();
+        }
+
+        public async Task PerformOperationsAsync(CancellationToken cancellationToken)
+        {
+            // Set handler to receive and respond to writable property update requests.
+            _logger.LogDebug("Subscribe to writable property updates.");
+            await _deviceClient.SubscribeToWritablePropertiesEventAsync(HandlePropertyUpdatesAsync, null, cancellationToken);
```

Comment:
Can we not have writable properties on the root level? Just wondering if the example shows that we are only handling writable properties on the components.

Reply:
We can, for sure, but the DTDL model that this sample uses does not define any root-level writable property for TemperatureController. We have a task for enhancing these DTDL models to include more scenarios, but we'll take it up post preview.